designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,338 @@
+-   snippets to produce typical (or starter) `aws-sam` debug configurations
+-   when no launch.json file is present in a workspace, VS Code exposes functionality that allows users to request auto-generated Debug Configurations. In this situation, the toolkit generates an `aws-sam` Debug Configuration for all `AWS::Serverless::Function` resources detected within all SAM Templates located in the workspace.
+
+Debug Configurations support local debugging for both [SAM Template resources](#sam-template-resource-local) and [standalone Lambda handlers](#standalone-lambda). Support for these different features is provided using different `request` field values in an `aws-sam` debug configuration.
+
+SAM Template resources are configured by setting `request` to `template-invoke`. A SAM Template and resource are required. This debug configuration uses the SAM Template as-is to debug as discussed above.
+
+Standalone Lambda handlers are configured by setting `request` to `standalone-lambda`. At a minimum, these configurations need to know what the Lambda handler is, which path is considered the project root, and what the Lambda runtime is. When this configuration is launched, a temporary SAM Template is produced. The template has a resource which references the Lambda handler. This temporary template (and resource) is then handled the same way as `template-invoke` configurations are.
+
+Example Debug Configuration entries can be found in the [Appendix](#sample-debug-configurations).
+
+### <a id="codelenses"></a> CodeLenses
+
+#### CodeLenses in SAM Template files
+
+CodeLenses are added to SAM Template files that serve as shortcuts to `template-invoke` [debug configurations](#debug-configurations) defined in the workspace.
+
+Every Debug Configuration that references a SAM Template and resource pairing will produce a CodeLens above that resource. When clicked, these CodeLenses launch the corresponding debug session, as if the user selected that debug configuration and pressed Debug from VS Code's Debug view.
+
+An additional CodeLens is placed above every template resource of type `AWS::Serverless::Function` called "Add Debug Configuration". Clicking this CodeLens produces a pre-filled `aws-sam` Debug Configuration in `launch.json` capable of performing a [local invoke](#sam-template-resource-local) of the associated resource.
+
+#### CodeLenses in Code files
+
+CodeLenses in code files assist with debugging standalone Lambda function handlers.
+
+A "Create Debug Configuration" CodeLens appears over functions considered eligible Lambda handlers. Clicking this CodeLens produces a pre-filled `aws-sam` Debug Configuration in `launch.json` capable of locally debugging a [standalone Lambda](#standalone-lambda), and opens `launch.json`.
```

Comment:
What is the eligibility condition? How likely are you to get code lenses for code that doesn't relate to AWS Lambda?

Reply:
I have added eligibility condition to the Appendix in the next revision.
It is likely code lenses will appear too often. I think we'll want to see how this initial experience works before refining that further.